c10/core/MobileAllocator.cpp

```diff
@@ -0,0 +1,93 @@
+#include <c10/core/MobileAllocator.h>
+#include <c10/core/CPUAllocator.h>
+
+namespace c10 {
+namespace {
+
+struct Configuration final {
+#if defined(__arm__) || defined(__aarch64__) || defined(_M_ARM) || defined(_M_ARM64)
+  static constexpr uint32_t kAlignment = 16u;
+#else
+  static constexpr uint32_t kAlignment = gAlignment;
+#endif
+};
+
+// QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors.
+// This behavior will trigger ASAN, and may result in a segfault if the accessed
+// memory location just so happens to fall on a page the current process has no
+// read access to.  Here we define a custom allocator that allocates the extra
+// storage required to keep this behavior safe.  This allocator could have been
+// restricted to QNNPACK and XNNPACK only, but that would have negative
+// performance ramifications, as input tensors must be reallocated, and copied
+// over, if not allocated with this allocator to begin with.  Making this
+// allocator the default on mobile builds minimizes the probability of unnecessary
+// reallocations and copies.
+//
+//
+// PreGuardBytes: Number of guard bytes to allocate before the allocation.
+// PostGuardBytes: Number of guard bytes to allocate after the allocation.
+
+template <uint32_t PreGuardBytes, uint32_t PostGuardBytes>
+class DefaultMobileCPUAllocator final : public at::Allocator {
+ public:
+  DefaultMobileCPUAllocator() = default;
+  virtual ~DefaultMobileCPUAllocator() override = default;
+
+  static void deleter(void* const pointer) {
+    const Cast memory{pointer};
+    c10::free_cpu(memory.as_byte_ptr - PreGuardBytes);
+  }
+
+  virtual DataPtr allocate(const size_t nbytes) const override {
+    Cast memory{
+      c10::alloc_cpu(
+        PreGuardBytes + nbytes + PostGuardBytes,
+        Configuration::kAlignment),
+    };
+
+    memory.as_byte_ptr += PreGuardBytes;
+
+    return {
+      memory.as_void_ptr,
+      memory.as_void_ptr,
+      &deleter,
+      at::Device(DeviceType::CPU),
+    };
+  }
+
+  virtual DeleterFnPtr raw_deleter() const override {
+    return deleter;
+  }
+
+ private:
+  union Cast final {
+    void * const as_void_ptr;
+    uint8_t * as_byte_ptr;
+  };
+};
+
+DefaultMobileCPUAllocator<std::max(8u, Configuration::kAlignment), 16u> g_mobile_cpu_allocator;
```

Comment:
QNNPACK requires 8 bytes guard before while XNNPACK requires 16 bytes after.  The thing to keep in mind here is that memory needs to be aligned not at the point of allocation, but 8 bytes into the allocation because that's what the kernels see.  Basically we need the equivalent of (Microsoft's) [_aligned_offset_malloc](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-offset-malloc?view=vs-2019) on all platforms, which is not present so must emulate.

This means we have to paths moving forward: 1) either make the pre-padding equal to the alignment so when we step that much forward into the buffer we land on an aligned location, or 2) use malloc to achieve the same behavior as a result of allocating extra "offset + pre_guard" bytes and doing pointer arithmatic to return an aligned address.  I chose the former but we can do the latter.

Reply:
This seems like a reasonable choice.  Can you leave a comment documenting where the constants 8 and 16 come from, and why we're doing the max?  You don't need to include the full context about MS's version and the alternative implementation.

Reply:
Done. Marking as resolved.